app/flatpak-builtins-utils.c

```diff
@@ -1435,15 +1437,19 @@ ensure_remote_state_all_arches (FlatpakDir         *dir,
                                 GCancellable       *cancellable,
                                 GError            **error)
 {
-  if (state->index_ht == NULL)
+  if (only_sideloaded)
     return TRUE;
 
-  GLNX_HASH_TABLE_FOREACH (state->index_ht, const char *, arch)
+  if (cached)
     {
-      if (!ensure_remote_state_arch (dir, state, arch,
-                                     cached, only_sideloaded,
-                                     cancellable, error))
+      /* First try cached, this will not error on uncached arches */
```

Comment:
I don't see how this is necessary. Doesn't the `flatpak_remote_state_ensure_subsummary_all_arches()` call below with `only_cached == FALSE` still prefer cached subsummaries?

Reply:
No. It is needed to keep the old behaviour.

The previous call to ensure_remote_state_arch(), when cached == TRUE, will first call ensure_subsummary with only_cache=TRUE and then if NOT_CACHED error, will call again with FALSE. This is the behaviour of --cached, i.e. prefer *any* local cache, even if its stale, but if you have no info at all, download it.

If you call flatpak_remote_state_ensure_subsummary_all_arches() with only_cached == FALSE, it will call ensure_subsummary() for all arches with only_cached = TRUE, which will *always* re-fetch the subsummary if the current local copy is stale. If you pass TRUE, it will instead load the ones that are locally available (even if stale). So, to get the old behaviour we first load all the locally available (TRUE) then call it again to fill out the required arches that failed and was left empty in the first call.